components/collector/src/metric_collectors/change_failure_rate.py

```diff
@@ -0,0 +1,78 @@
+"""Change failure rate collector."""
+
+from collections.abc import Sequence
+from datetime import datetime
+from typing import ClassVar
+
+from base_collectors import MetricCollector
+from collector_utilities.date_time import parse_datetime
+from model import Entities, Entity, MetricMeasurement, SourceMeasurement
+
+
+class ChangeFailureRate(MetricCollector):
+    """Change failure rate collector."""
+
+    # Partial sources for issues and deployments, which need to be combined within this metric
+    ISSUE_SOURCE_TYPES: ClassVar[list[str]] = ["jira"]
+    DEPLOYMENT_SOURCE_TYPES: ClassVar[list[str]] = ["gitlab", "jenkins"]
+
+    async def collect(self) -> MetricMeasurement | None:
+        """Override to calculate the composite value."""
+        if (measurement := await super().collect()) is None:
+            return None
+        deployments = self.deployment_entities(measurement.sources)
+        deployments.sort(key=lambda d: self.deployment_timestamp(d), reverse=True)  # newest to oldest
+        issues = self.issue_entities(measurement.sources)
+        for deployment in deployments:
+            issue_length = len(issues)
+            if issue_length == 0:
+                break  # no more issues; there are no more failed deployments
+            issues = Entities(
+                issue for issue in issues if self.deployment_timestamp(deployment) > self.issue_timestamp(issue)
+            )
+            if len(issues) != issue_length:
+                deployment["failed"] = True  # there were issues after this deployment, so it counts as failed
```

Comment:
The `deployments` are sorted from newest to oldest (line 24), `issues` are only those for which deploy dt > issue dt (line 31). The loop iterates backwards so it can deal with the open ended interval `[deploy_dt -> deploy_dt | inf)`